Beta Version of Blender Exporter

ProjectBarks opened the pull request:
The design of the exporter was to incorporate blender design paradigms while working within available tooling.

# Design
Files follow standard python guidelines in code seperation while placing a larger emphasis on smaller readable files rather than large unreadable files. Files placed in `operators` are executed based tasks that influence the user interface...
- `oauth` starts a server to catch the token and shuts down once received
- `token` once the token is received caching occurs while forcing a update
- `upoad_gltf` exports a base gltf and uploads it after receiving required tokens


The `panels` are minimal implementation of UI used primarily as a user facing endpoint. `__init__` handles registration and loading preferences. `cache` is a simple way of dumping preferences into json. `server` is a dumb http server that returns `index.html` and shuts down on the first token it receives. All tokens and basic settings are stored in `globals.py`


**IGNORE VENDORS** - It is required dependencies for blender but not automatically installed

# Todo
- Add Menu File export
- Add Export Options

ProjectBarks committed:
Alpha version

ProjectBarks committed:
Update todo

ProjectBarks committed:
Finish merge

ProjectBarks committed:
Add gitignore

ProjectBarks committed:
remove ds store

ProjectBarks committed:
add installation instructions

ProjectBarks committed:
format code

ProjectBarks committed:
Code seperation

ProjectBarks committed:
Added token caching

ProjectBarks committed:
update readme to reflect changes

ProjectBarks committed:
[Security] Leaking auth token fix & [Feature] Add progress

Using context.scene would export a blender file with the users auth_token saved inside if this file were to be shared someone could easily have write access to an ion account

Progress bar has been added to the upload process although it has yet to be seen locally

ProjectBarks committed:
Add ability to upload from file options

Go into File > Export > Upload to Ion and all glb files are automatically uploaded with default settings to Cesium Ion

mramato commented on README.md:
```diff
@@ -1,2 +1,15 @@
+
+## Installation
+1. Download the project as a zip [here](https://github.com/AnalyticalGraphicsInc/ion-blender-exporter/archive/beta.zip)
+2. Go to `File > Preferences > Add-ons`
```
I had to go to `Edit -> Preferences -> Add-ons`.

mramato commented on README.md:
```diff
@@ -1,2 +1,15 @@
 Blender add-on for uploading and tiling models with Cesium ion. https://cesium.com
+
+## Installation
+1. Download the project as a zip [here](https://github.com/AnalyticalGraphicsInc/ion-blender-exporter/archive/beta.zip)
```
This is OK for now, but we'll need to figure out the release process, I assume we will eventually link to the latest official releases taged with GitHub releases.

ProjectBarks replied:
Yep

ProjectBarks replied:
TODO: Switch this to using releases instead & Write release documentation

mramato commented on README.md:
```diff
@@ -1,2 +1,15 @@
+2. Go to `File > Preferences > Add-ons`
+3. Enable Add-on
+
+## Todo
```
We don't keep TODO lists in the README.  You can move these into the PR description for now (we normally use task lists for this kind of stuff).  For stuff not part of the PR, we should write up new issues.

ProjectBarks replied:
Yeah I thought it would help contextualize where I am taking the codebase

mramato replied:
That's a valid approach, just not how we use READMEs throughout Cesium. At long as it's cleaned up before we merge, that's fine.

mramato commented on .gitignore:
```diff
@@ -0,0 +1,130 @@
+###OSX###
```
We use minimal ignore lists to make them easier to maintain.  Please remove everything here that isn't necessary for this project.

mramato commented on __init__.py:
```diff
@@ -0,0 +1,48 @@
+bl_info = {
+    "name": "Cesium Ion",
+    "description": "Cesium Ion Exporting Tool",
+    "author": "",
```
Author should be `Analytical Graphics, Inc.`

mramato commented:
Haven't looked at the code yet, but here's some general top-level UX feedback from my first few minutes (some of these we may have already discussed)

1. Why have a separate `Authorize` step, can't we just authorize the first time they hit the export button?
2. Rather than have a modeless dialog that they have to fill in, why not a modal that pops up with whatever options are available (this will also make it easy to pre-populate it with defaults from the Blender file we are exporting.
3. Similar to 2, from this tutorial https://docs.blender.org/manual/en/dev/advanced/scripting/addon_tutorial.html, it looks like we could just either insert ourselves in the `File` menu, or add a new top-level Cesium ion menu (though I think it makes more sense to `File->Export->Cesium ion` or `File->Upload-> Cesium ion` if we can't add to the sub-menu.
4. We need to have a combobox or radio button that let's uses choose between `3D Model` and `3D Capture` so that we can set the equivalent option in the REST API (there shouldn't be a default, make them choose).

mramato commented:
Is `vendor` required to be the directory name or otherwise standardized in python?  In all Cesium codebases, we always call this `ThirdParty`

mramato commented on __init__.py:
```diff
@@ -0,0 +1,48 @@
+bl_info = {
+    "name": "Cesium Ion",
```
`ion` is always lower-cased.  The only exception is code symbols for readability `IonClass`, `CesiumIonClass`, etc..

mramato commented on __init__.py:
```diff
@@ -0,0 +1,48 @@
+bl_info = {
+    "name": "Cesium Ion",
+    "description": "Cesium Ion Exporting Tool",
```
Perhaps something like `Upload and tile models with Cesium ion. https://cesium.com` We always have to be mindful when communicating information to the user.  I assume this is just plain text?  (No markdown or anything?)

mramato commented:
This is coming along nicely.  Great work @ProjectBarks! Since I asked for some non-trivial changes I'm going to stop reviewing for now.  Let me know if you have any additional questions and just bump when you think it's ready for another look.

ProjectBarks committed:
Renaming ion references & Adding User Preferences

Added alternative form of handling preferences in Edit > Add-on > Cesium ion

Renamed Ion to ion

ProjectBarks committed:
Beta Merge Requests

Changed readme wording

Reduced size of .gitignore

Reworded bl_info

Removed file menu

made authorization smoother

renamed vendor directory

Added texture options & WEBP exporting

ProjectBarks commented:
Ready for review

ProjectBarks commented:
@mramato ready for review

ProjectBarks committed:
Update ReadMe & Fix label bug

Better installation instructions were provivded that give detailed instructions on how to use

There was a bug where certain labels had double colons like 'Name::'

ProjectBarks committed:
Fix padding

padding caused pushover on github.com

mramato commented:
Taking a look now.

ProjectBarks committed:
Add author

mramato commented on panels/export.py:
```diff
@@ -0,0 +1,69 @@
+
+            layout.prop(scene.csm_export, "name")
+            if len(csm_export.name) <= 0:
+                layout.label(text="Set name before export...")
```
Let's just change this label to `Name is required`.

mramato commented on panels/export.py:
```diff
@@ -0,0 +1,69 @@
+            layout.separator()
+            layout.prop(scene.csm_export, "source_type")
+            if csm_export.source_type is "":
+                layout.label(text="Set source type before export...")
```
For UX purposes, let's refer to it as `Model type` (since Blender only works in models while source is more general to ion's multi-domain usage).  Also change this to `Model type is required`.

mramato commented:
This turned out super nice, awesome work @ProjectBarks.  I think I just have some minor stuff that I'm going through now.

ProjectBarks committed:
Small naming changes

ProjectBarks commented:
@mramato made those naming changes

mramato commented on panels/export.py:
```diff
@@ -0,0 +1,69 @@
+SOURCE_TYPES = [
+    ("3D_MODEL", "3D Model", "Models designed through blender", "FILE_3D", 1),
+    ("3D_CAPTURE", "3D Capture",
+     "Model created through photogrammetry processes", "CAMERA_DATA", 2)
```
> Model created through photogrammetry processes

Just realized the doc in the REST API is pretty bad.  Let's change this to something like:

Model created through 3D capture such as scanning, LIDAR, photogrammetry, etc...

mramato commented:
I'm guessing the answer is no, but is there way to expand the default width of the tab?

mramato commented:
Can we show some sort of feedback if they click `Upload to ion` without entering a name?  Mobile apps shake, or maybe hide the required messages until there is an actual error?  What does Blender's UX allow for?

mramato commented on panels/export.py:
```diff
@@ -0,0 +1,69 @@
+                              description="Type of data being uploaded",
+                              items=lambda self, context: SOURCE_TYPES)
+    webp_textures: BoolProperty(
+        name="Convert to WebP",
```
I think this is confusing since it doesn't say what will be WebP.  `Use WebP images` is what the ion UX has, so let's go with that.

mramato commented on operators/upload_gltf.py:
```diff
@@ -0,0 +1,171 @@
+class ExportUploadOperator(Operator):
+    bl_label = "gltf to Cesium uploader"
+    bl_idname = f"{APP_OPERATOR_PREFIX}.upload_gltf"
+    bl_description = "Uploads \".glb\" exports to Cesium ion"
```
This is oddly phrased.  Let's go with `Uploads your model to Cesium ion for 3D Tiling`

ProjectBarks commented:
> I'm guessing the answer is no, but is there way to expand the default width of the tab?
You guess correctly

> Can we show some sort of feedback if they click `Upload to ion` without entering a name? Mobile apps shake, or maybe hide the required messages until there is an actual error? What does Blender's UX allow for?
I changed the UI so it is more obvious of what is required. The best we can do is show a required error icon for things not complete. Checkout the newest revision

ProjectBarks committed:
Change labels

ProjectBarks committed:
Disable local development & add external parameters

ProjectBarks committed:
undo upload icon

ProjectBarks committed:
fix ion addres

mramato commented:
Awesome work @ProjectBarks.
